[CD] Add Python Docker Pipeline

perdasilva opened the pull request:
* Refactors docker login functionality out of ci/docker_cache.py to make it generic
* Refactors docker run functionality out of ci/build.py to make it general purpose
* Updates CD pipeline for producing python docker images

perdasilva committed:
Adds python docker pipeline

perdasilva committed:
Refactors docker login functionality out of docker_cache script

tweak

perdasilva committed:
Refactors docker run functionality out of build.py

Test safe docker run tweaks

Fix up

perdasilva committed:
Adds CD docker utils

perdasilva commented on cd/python/docker/python_images.sh:
```diff
@@ -0,0 +1,130 @@
+    ./ci/safe_docker_run.py ${runtime_param} --cap-add "SYS_PTRACE" -u `id -u`:`id -g` -v `pwd`:/work/mxnet "${test_image_name}" ${resources_path}/test_python_image.sh "${mxnet_variant}" "${python_cmd}"
+}
+
+# Pushes the runtime image to the repository
```
Comments don't seem related to this docker image...

perdasilva commented on ci/docker_cache.py:
```diff
@@ -24,21 +24,21 @@
-DOCKER_CACHE_NUM_RETRIES = 3
 DOCKER_CACHE_TIMEOUT_MINS = 45
 PARALLEL_BUILDS = 10
+DOCKER_CACHE_NUM_RETRIES = 3
```
Move this to the top to minimize change set

perdasilva commented on ci/docker_login.py:
```diff
@@ -0,0 +1,137 @@
+    logging.info(p.stdout)
+    if p.returncode == 0:
+        logging.info('Successfully logged in to DockerHub')
+        return True
```
No need to return true here - failure results in Error getting raised

perdasilva commented on cd/python/docker/Jenkins_pipeline.groovy:
```diff
@@ -0,0 +1,74 @@
+def build(mxnet_variant) {
+  ws("workspace/python_docker/${mxnet_variant}/${env.BUILD_NUMBER}") {
+    ci_utils.init_git()
+    cd_utils.restore_artifact(mxnet_variant, 'static')
```
Maybe create two variants of this method one for static and one for dynamic, instead of passing this in as a parameter. Then, call the methods restore_static_libmxnet and restore_dynamic_libmxnet

perdasilva commented on ci/test_safe_docker_run.py:
```diff
@@ -0,0 +1,396 @@
+            safe_docker = SafeDockerClient()
+            assert safe_docker.run("image", "command") == 152
+
+    def test_volume_mount(self):
```
Maybe this should be an internal method and the translation should be checked in the main function tests

perdasilva commented on ci/safe_docker_run.py:
```diff
@@ -0,0 +1,227 @@
+    parser.add_argument("command", metavar="COMMAND")
+    parser.add_argument("args", nargs='*', metavar="ARG")
+
+    args = parser.parse_args(args=command_line_arguments[1:])
```
maybe pass in sys.argv[1:] instead of doing command_line_arguments[1:] here...

perdasilva commented on ci/docker_login.py:
```diff
@@ -0,0 +1,137 @@
+                        type=str,
+                        default=os.environ.get("DOCKERHUB_SECRET_ENDPOINT_REGION", None))
+
+    args = parser.parse_args(args=sys_argv[1:])
```
Maybe pass in sys.argv[1:] to the main method, instead of doing sys_argv[1:] here

larroy commented on ci/docker_login.py:
```diff
@@ -0,0 +1,137 @@
+        description="Safe docker login utility to avoid leaking passwords",
+        epilog=""
+    )
+    parser.add_argument("--secret-name",
```
Beautiful, if we could ask for the cherry on top of the cake, being able to pass user and password in case we don't want / we can't use secrets manager it would be great.

perdasilva replied:
I won't put this through because we had a whole CoE on leaked docker passwords - this is why we ended up having to use the secrets manager =P I'm sure if I add it, Marco will flip... ._.

larroy commented on ci/safe_docker_run.py:
```diff
@@ -0,0 +1,227 @@
+        return ret
+
+
+def volume_mount(volume_dfn: str) -> Dict[str, Any]:
```
Nice!  for icing on the cake adding a docstring :-)

larroy commented on ci/test_safe_docker_run.py:
```diff
@@ -0,0 +1,396 @@
+#!/usr/bin/env python3
```
This is great

larroy commented on ci/util.py:
```diff
@@ -113,3 +116,14 @@ def chdir_to_script_directory():
+    return os.path.split(sys.argv[0])[1]
+
+
+def config_logging():
```
Could we update this with something like:

```

def config_logging():
    logging.config.fileConfig(os.getenv('LOGGING_CONF', 'logging.conf'))
```


```
[loggers]
keys=root,simpleExample

[handlers]
keys=consoleHandler

[formatters]
keys=simpleFormatter

[logger_root]
level=DEBUG
handlers=consoleHandler

[logger_simpleExample]
level=DEBUG
handlers=consoleHandler
qualname=simpleExample
propagate=0

[handler_consoleHandler]
class=StreamHandler
level=DEBUG
formatter=simpleFormatter
args=(sys.stdout,)

[formatter_simpleFormatter]
format=%(asctime)s - %(name)s - %(levelname)s - %(message)s
datefmt=
```

I'm moving away from these hardcoded logging statements.

larroy reviewed (changes requested):
Great work @perdasilva thanks a lot for this. 

Everything looks quite good in general, just the     jenkins_env_vars = ['BUILD_NUMBER', 'BUILD_ID', 'BUILD_TAG']   I don't see them set. these have to be set and percolated through the containers so Jenkins can kill runaway containers as the signal is not always sent / there's no time to act when the jobs are interrupted.

larroy commented:
Also, great stuff with the tests and mocking, great to learn from you.

perdasilva committed:
Updates logging configuration to file config

perdasilva commented:
@larroy thank you so much for taking the time to review - it was specially important to get the zombie guarding right - I will put through the the env vars thing through and add the tests. This way it is documented and ensured! Thanks again ^^

perdasilva committed:
Adds restore static and dynamic methods, instead of making the library type a parameter

perdasilva committed:
Removes CUDA 9.1 from base images script

perdasilva committed:
Adds explanatory comment to signal masking in safe_docker_run

perdasilva committed:
Adds push success message with pull command and link to repository

perdasilva committed:
Fixes logging

larroy commented:
I will help review this again in the next couple of days. Feel free to ping me if I'm not responsibe as I missed your latest changes. Thanks again for following up on this and let's try to get it merged so we can open smaller PRs on top.

perdasilva commented:
@larroy - after addressing your comments I created a PR against incubator - it's been merged sans issues thanks to you ^^ we are now producing nightly python images for mxnet XD
